### ISSUE TITLE
Asserting boolean variable fails, despite successfully asserting its definition

### ISSUE THREAD

**Issue by ArmborstL** [viperproject/silicon#483]
Dear Viper team,
thank you for your work!
I am newly working on the VerCors front-end for Viper. I stumbled upon the following issue:
``` 
var b: Bool
assert A ==> B    // passes
b := A==>B
assert A ==> B    // passes
assert b          // fails 
```
```A``` and ```B``` are somewhat more complex expressions, involving functions and quantifiers. Nevertheless, I would expect the last assert to hold.
It might be related to a lack of triggers on the quantifier. But I am confused why it then can assert the exact same expression directly.
I attach a file containing the full example. It is generated code, so somewhat convoluted. My apologies for that. I tried to simplify it a little, but I could not remove too much without making the assertion pass (e.g. turning the ```Array[Ref]``` parameters into ```Array[Int]``` makes it pass). The relevant part is line 36-39.
[viper.txt](https://github.com/viperproject/silicon/files/4360223/viper.txt)

Thank you for looking into this :)

PS: This is on Viper Release 2020.1. Let me know if you need more information.

**Comment by mschwerhoff** [comment]
I've quickly tried Carbon, the other Viper back-end, and it verifies the example. So it indeed seems to be an issue of Silicon.

Just to double-check: does the program verify if you remove the final `assert b`?

The file you attached (`viper.txt`) looks non-trivial at first glance. I know you've already cut it down a bit, but any further simplification would be much appreciated. Options for that include inlining predicate or function bodies, removing conjunctions, and killing of branches (from implications etc.) by additional assumptions.

**Comment by ArmborstL** [comment]
Yes, without the final assertion, the program verifies.

Unfortunately, I did not manage to simplify it much further without losing the error. I could inline one function, but inlining the second made the program pass. I also moved some boundary conditions into function preconditions to simplify the body. I could not remove them entirely, as they are boundary conditions to make the following expression(s) well-formed. I attached the resulting file [[viper2.txt](https://github.com/viperproject/silicon/files/4374878/viper2.txt)]. It does not look simpler, but maybe it is to Viper. Now, it is basically
```
function fun() : Bool { expr1 }
var b: Bool
var r: Bool
b := fun() 
r := (b ==> expr1)
assert r          // fails unexpectedly
```
with the expression ```expr1``` being a ```forall```. The interesting assertions are still Lines 36-39.
In the process of shuffling things around, I noticed that if I add any conjunction to the function call, the verification passes. In particular, turning ```reordered := IsReordered(<...>)``` into ```reordered := true&&IsReordered(<...>)``` solves the issue (see Line 33). This is obviously _highly unintuitive_ behaviour, making the issue difficult to debug.
Thanks for your support!

**Comment by fpoli** [comment]
I tried to simplify the example:

```
domain Array[T] {
	function loc(a: Array[T], i: Int): T
	function len(a: Array[T]): Int
}

field val: Int

define read_permissions(array) (
	forall i: Int :: {loc(array, i)} 0 <= i && i < len(array) ==> acc(loc(array, i).val, 1/2)
)

define property(a, b) (
	forall i: Int :: {loc(b, i)}
		0 <= i && i < len(a) && i < len(b) ==> loc(b, i).val == loc(a, i).val
)

function Property(a: Array[Ref], b: Array[Ref]): Bool
	requires read_permissions(a)
	requires read_permissions(b)
{
	property(a, b)
}

method Test(a: Array[Ref], b: Array[Ref])
	requires read_permissions(a)
	requires read_permissions(b)
{
	assert Property(a, b) == property(a, b)   // Ok
	assert Property(a, b) ==> property(a, b)  // Error
}
```

**Comment by mschwerhoff** [comment]
@fpoli Thank you!

**Comment by mschwerhoff** [comment]
More madness:

```
assert Property(a, b) == property(a, b)

assume Property(a, b)
assert property(a, b)
```

The final assert fails, but if the first assertion is removed, it holds.

**Comment by mschwerhoff** [comment]
A version that (always) fails in Silicon:

```
domain Array[T] {
	function loc(a: Array[T], i: Int): T
}

field val: Bool

define read_permissions(array) (
	forall i: Int :: {loc(array, i)}
		0 <= i ==> acc(loc(array, i).val, 1/2)
)

define property(b) (
	forall i: Int :: {loc(b, i)}
		0 <= i ==> loc(b, i).val
)

function Property(a: Array[Ref], b: Array[Ref]): Bool
	requires read_permissions(a)
	requires read_permissions(b)
{
	property(b)
}

method Test(a: Array[Ref], b: Array[Ref])
{
	inhale read_permissions(a)
	inhale read_permissions(b)

  var tmp1: Bool := Property(a, b)
  var tmp2: Bool := property(b)

  assert tmp1 == tmp2 // Fails in Silicon
}
```

**Comment by mschwerhoff** [comment]
Resolved with commit d7ffe91: An erroneously returned result from the snapshot map cache resulted in the observed incompleteness.

@ArmborstL Two remarks:
* Don't expect *intuitive behaviour* from bugs. There is no logic to them — at least not beyond the technical root cause.
* I've noticed that all quantifiers in the example lack triggers. While Viper does a best-effort attempt to infer triggers, it is in general strongly advised to explicitly specify them.

Please don't hesitate to ask, if you have any questions.
